langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/VertexAiGeminiChatModel.java

```diff
@@ -101,16 +99,51 @@ public VertexAiGeminiChatModel(GenerativeModel generativeModel,
 
     @Override
     public Response<AiMessage> generate(List<ChatMessage> messages) {
-
         List<Content> contents = ContentsMapper.map(messages);
-
         GenerateContentResponse response = withRetry(() -> generativeModel.generateContent(contents, generationConfig), maxRetries);
 
-        return Response.from(
+        Content content = ResponseHandler.getContent(response);
+        Optional<Part> containsFunctionCall = content.getPartsList().stream().filter(Part::hasFunctionCall).findFirst();
+
+        if (containsFunctionCall.isPresent()) {
+            FunctionCall functionCall = containsFunctionCall.get().getFunctionCall();
+            ToolExecutionRequest toolExecutionRequest = FunctionCallHelper.fromFunctionCall(functionCall);
+
+            return Response.from(
+                AiMessage.from(toolExecutionRequest),
+                TokenUsageMapper.map(response.getUsageMetadata()),
+                FinishReasonMapper.map(ResponseHandler.getFinishReason(response))
+            );
+        } else {
+            return Response.from(
                 AiMessage.from(ResponseHandler.getText(response)),
                 TokenUsageMapper.map(response.getUsageMetadata()),
                 FinishReasonMapper.map(ResponseHandler.getFinishReason(response))
-        );
+            );
+        }
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
+        try {
+            Tool tool = FunctionCallHelper.convertToolSpecifications(toolSpecifications);
+            //TODO: hack for now, till the next release that will have a tool param
+            // in the underlying generateContent() method
+            this.generativeModel.setTools(Collections.singletonList(tool));
+            return generate(messages);
+        } finally {
+            //TODO: remove the finally block (see above todo)
+            this.generativeModel.setTools(Collections.emptyList());
```

Comment:
Should we create a new instance of `generativeModel` in this method for the duration of this call instead of sharing the same instance and hope there will be no concurrent calls?

Reply:
Yes, that's another approach which is safer.
But I'm waiting for the new Gemini SDK to be released (hopefully at the end of the week), and I'll be able to remove this `try`/`finally` block, and thus have a cleaner solution.

Reply:
Makes sense!

FYI, I plan to do a release tomorrow because there are some long-awaited features and bugs pending to be released.
And probably another one on 15 or 22.
So it is up to you if you want to release it tomorrow and then refactor (only internal changes) or wait for a proper solution

Reply:
Ok, thanks for the heads up.
I'll try to work on it today / tonight, and I'll update you to see how far I am.